### PR TITLE
Implements screen modes

### LINKED ISSUE #2327: Implement mode concept
The `switch_screen` method doesn't fit with the stack of screens concept.

It was introduced to provide switchable modes in an app, but doesn't quite capture that. I think what we need is a concept of "modes" where each mode has it's own independant stack of screens. The app can switch between these modes as required.

Suggest we have a `MODES` class var similar to screens, which should be a sequence of modes.

Something like this:

```python
MODES = ["calculator", "help", "error", "sign-up"]
```

Each of those modes has an associated stack of screens.

App would have a `switch_mode` method which changes to the given mode. Modes could also be created dynamically, with `add_mode` and `remove_mode`.

I'm hoping this could be implemented without much impact on the rest of the code. But there may be edge cases.

If you tackle this, please check with @willmcgugan

### PR DESCRIPTION
**Pull request by rodrigogiraoserrao**
As it stands, the modes "work" but the API feels a bit rough.
Discussion in the related issue: #2327

The app below lets you play with modes.
Press <kbd>1</kbd> and <kbd>2</kbd> to switch between modes 1 and 2.
Press <kbd>P</kbd> to push random fruit modals into the screen.
Press <kbd>O</kbd> to pop a screen from the stack.

<details>
<summary>Demo code.</summary>

```py
from random import choice

from textual.app import App, ComposeResult
from textual.screen import Screen, ModalScreen
from textual.widgets import Footer, Header, Label


FRUITS = "apple mango strawberry banana peach pear melon watermelon".split()


class BaseScreen(Screen[None]):
    BINDINGS = [
        ("1", "one", "Mode 1"),
        ("2", "two", "Mode 2"),
        ("p", "push", "Push rnd scrn"),
        ("o", "pop_screen", "Pop"),
        ("r", "remove", "Remove mode 1"),
    ]

    def __init__(self, label):
        super().__init__()
        self.label = label

    def compose(self) -> ComposeResult:
        yield Header()
        yield Label(self.label)
        yield Footer()

    def action_one(self) -> None:
        self.app.switch_mode("one")

    def action_two(self) -> None:
        self.app.switch_mode("two")

    def action_push(self) -> None:
        self.app.push_screen(FruitModal())

    def action_remove(self) -> None:
        self.app.remove_mode("one")


class FruitModal(ModalScreen[None]):
    BINDINGS = [
        ("1", "one", "Mode 1"),
        ("2", "two", "Mode 2"),
        ("p", "push", "Push rnd scrn"),
        ("o", "pop_screen", "Pop"),
    ]

    def compose(self) -> ComposeResult:
        yield Label(choice(FRUITS))

    def action_one(self) -> None:
        self.app.switch_mode("one")

    def action_two(self) -> None:
        self.app.switch_mode("two")

    def action_push(self) -> None:
        self.app.push_screen(FruitModal())


class ModesApp(App[None]):
    MODES = ["one", "two"]

    def on_mount(self) -> None:
        self.switch_mode("two")
        self.push_screen(BaseScreen("two"))
        self.switch_mode("one")
        self.push_screen(BaseScreen("one"))


app = ModesApp()
if __name__ == "__main__":
    app.run()
```

</details>

This will close #2327

### CODE REVIEW

**Comment by rodrigogiraoserrao** [comment]
I like this API more.
User specifies modes and base screens for all modes:

```py
class HelpScreen(Screen[None]):
    ...

class MainAppScreen(Screen[None]):
    ...

class MyApp(App[None]):
    MODES = {
        "default": "main",
        "help": HelpScreen,
    }

    SCREENS = {
        "main": MainAppScreen,
    }

    ...
```

Then, inside `on_mount` the user can immediately switch to a desired mode.

`add_mode` now takes a mode name and the base screen for that mode.

**Comment by willmcgugan** [comment]
@rodrigogiraoserrao Where are we with this? Is it in progress or are you blocked?

File: tests/test_screen_modes.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+        # Check the stack.
+        assert len(app.screen_stack) == N + 1
+        for _ in range(N):
+            assert str(app.query_one(Label).renderable) == fruits.pop()
+            await pilot.press("o")
+
+
+async def test_inactive_stack_is_alive():
+    """This tests that timers in screens outside the active stack keep going."""
```
Comment (rodrigogiraoserrao) [inline tests/test_screen_modes.py:177]: Please, someone explicitly confirm this is the intended behaviour / desirable behaviour.

File: src/textual/app.py [hunk 2]
```diff
@@ -1326,6 +1376,89 @@ def mount_all(
...
+            del self.MODES[mode]
+
+        if mode not in self._screen_stacks:
+            return
+
+        stack = self._screen_stacks[mode]
+        for screen in reversed(stack):
+            if screen._running:
+                self.call_later(self._prune_node, screen)
+        del self._screen_stacks[mode]
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/app.py:1460]: I have a suspicion that it isn't. See the code in `remove()`. The screens and their widgets should get an `Unmount` message

File: src/textual/app.py [hunk 3]
```diff
@@ -1398,11 +1531,14 @@ def _replace_screen(self, screen: Screen) -> Screen:
...
         """
-        if self._screen_stack:
+        if self._screen_stacks[self._current_mode]:
             self.screen.refresh()
         screen.post_message(events.ScreenSuspend())
         self.log.system(f"{screen} SUSPENDED")
-        if not self.is_screen_installed(screen) and screen not in self._screen_stack:
+        if (
+            not self.is_screen_installed(screen)
+            and screen not in self._screen_stacks[self._current_mode]
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/app.py:1540]: I think you should check that it is not in any stack. You don't want to remove it, if it is in use.

File: tests/test_screen_modes.py (hunk 1 above)
Comment (willmcgugan) [reply tests/test_screen_modes.py:177]: Timers should be orthogonal to the state of the screen.

Comment (rodrigogiraoserrao) [reply tests/test_screen_modes.py:177]: In other words, a screen being moved out of the active stack should _not_ interfere with timers?

File: src/textual/app.py [hunk 4]
```diff
@@ -1326,6 +1376,89 @@ def mount_all(
...
+    def _init_mode(self, mode: str) -> None:
+        """Do internal initialisation of a new screen stack mode."""
+
+        stack = self._screen_stacks.get(mode, [])
+        if not stack:
+            _screen = self.MODES[mode]
+            if callable(_screen):
+                screen, _ = self._get_screen(_screen())
+            else:
+                screen, _ = self._get_screen(self.MODES[mode])
```
[the remark this answers is not in the document]
Comment (rodrigogiraoserrao) [reply src/textual/app.py:1388]: It seems to work as expected, yes!

File: src/textual/app.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rodrigogiraoserrao) [reply src/textual/app.py:1460]: Right. I redid this based on the code in `App._replace_screen`.

File: src/textual/app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rodrigogiraoserrao) [reply src/textual/app.py:1540]: Right, that makes sense.

**Review by willmcgugan — APPROVED** [review]
Looks good! just one request.

File: src/textual/app.py [hunk 5]
```diff
@@ -680,15 +728,17 @@ def screen(self) -> Screen:
             ScreenStackError: If there are no screens on the stack.
         """
         try:
-            return self._screen_stack[-1]
+            return self._screen_stacks[self._current_mode][-1]
```
[the remark this answers is not in the document]
Comment (rodrigogiraoserrao) [reply src/textual/app.py:740]: Did you mean to ask for a property for `self._screen_stacks[self._current_mode]`?

[the remark this answers is not in the document]
Comment (rodrigogiraoserrao) [reply src/textual/app.py:740]: There is already a `screen_stack` property that returns a copy of the current stack but in some places I want the actual stack.

[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/app.py]: How about a `_screen_stack` property then?

File: src/textual/app.py [hunk 6]
```diff
@@ -159,6 +159,38 @@ class ScreenStackError(ScreenError):
     """Raised when trying to manipulate the screen stack incorrectly."""
 
 
+class ModeError(Exception):
```
[the remark this answers is not in the document]
Comment (rodrigogiraoserrao) [reply src/textual/app.py:162]: Fixed in #2715.
What a weird thing, though...

**Pull request merged**